core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java

```diff
@@ -267,14 +284,48 @@ public void receiveState(int segment, Object key, Object lockOwner, IracMetadata
    }
 
    // public for testing purposes
-   public void sendStateIfNeeded(Address origin, IntSet segments, int segment, Object key, Object lockOwner) {
-      if (!segments.contains(segment)) {
-         return;
+   void transferStateTo(Address dst, IntSet segments, Collection<? extends IracManagerKeyState> stateCollection) {
+      if (log.isTraceEnabled()) {
+         log.tracef("Starting state transfer to %s. Segments=%s, %s keys to check", dst, segments, stateCollection.size());
+      }
+      Flowable.fromIterable(stateCollection)
+            .filter(s -> !s.isStateTransfer() && !s.isExpiration() && segments.contains(s.getSegment()))
+            .buffer(batchSize)
+            .concatMapCompletable(batch -> createAndSendBatch(dst, batch))
+            .subscribe(new CompletableObserver() {
+               @Override
+               public void onSubscribe(@NonNull Disposable d) {
+
+               }
+
+               @Override
+               public void onComplete() {
+                  if (log.isTraceEnabled()) {
+                     log.tracef("State transfer to %s finished!", dst);
+                  }
+               }
+
+               @Override
+               public void onError(@NonNull Throwable e) {
+                  if (log.isTraceEnabled()) {
+                     log.tracef(e, "State transfer to %s failed!", dst);
```

Comment:
How does `Flowable` behave in case of exception? Does an exception in one of the intermediate steps stop the iterate an `onError` is invoked?

In theory, I should not have any exception in the intermediate steps.

offtopic question: in which thread is the iteration running? should I set up to use (non)blocking executor?

Reply:
> How does Flowable behave in case of exception? Does an exception in one of the intermediate steps stop the iterate an onError is invoked?

It depends upon the operator used. RxJava has an option for eager exception handling or not. Looks like you want concatMapCompletableDelayError instead.

> offtopic question: in which thread is the iteration running? should I set up to use (non)blocking executor?

The flowable here will be subscribed to on the invoking thread, however once you hit the `concatMapCompletable` where it isn't completed the current thread will return and it will be resumed on whatever thread completes that Completable. You shouldn't have to worry about any executor/threads here though.

Reply:
changed to `concatMapCompletableDelayError`. thanks!